### ISSUE TITLE
Auto bed leveling grid on/off Z height difference

### ISSUE THREAD

**Issue by imqqmi** [MarlinFirmware/Marlin#1244]
Just found out after printing a 20x20x50mm empty cube that the Z height is different when I do only a G28 before printing or G28 + G29 before printing. The difference is about 3mm (46.54mm vs 49.54mm). I've repeated the test to be sure and indeed the only setting that was changed was the auto bed leveling. 

Is this a known bug? The firmware seems to be from before 14-11-2014 (dd-mm-yyy). I'm printing with a Prusa i3 and a RAMPS 1.4 board.

**Comment by imqqmi** [comment]
Still good news but my interpretation was wrong. I've swapped glass plates when I did my last test, I was going to print ABS on vinyl which has so far been the best adhesive that I've tried. The other plate was for hairspray and glue stick test. 

I thought the z axis nuts looseness were the cause. while it did slightly improve layer height as I can now more precisely do the first layer (was always about half the height it should've been). 

Apparently the way glass plate 1 with vinyl was shaped/bend/flexed made ABLG reduce the size of about 7%. At the moment I swapped the glass plate to #2 the height was really close to the target. 

I couldn't get the glass plates flatter and both had a bend over the diagonal (flexing, or torsion). I've now used a 1.5mm aluminium plate that I've shaped and bend until the probing indicated less than 0.07 deviation. Glass plate couldn't get below 0.25. Measured Z height during ABLG was between 0.46 and 0.61. 

Now it seems to work consistently after 4 ABS 20x20x25mm cubes with exactly the same height of 24.85mm. I'd say that's close enough for now, up from 23.30mm. 

The gist of the story is that the bed needs to be as flat as a pancake. It may be at an angle as a whole but the surface can't be bend or have torsion/flex shape or it may cause some height issues. 

I've also experimented with the number of clamps on the glass plate. 2 improved the flatness compared to 4 clamps. Of course the contact the the heatbed isn't as good so heat is not as even on the glass plate. That's also why I switched to aluminium, even if it doesn't have a great contact the heat transfer spreads the heat more evenly.

Here are the probings and M499 values (multiplied by 1000 so you can see the actual values). I took two G29 for each plate and number of clamps 2 or 4, there are 8 sets of values:

ABLG glass plate 2, 4 clamps

Bed x: 30.00 y: 50.00 z: -0.19
Bed x: 87.00 y: 50.00 z: -0.06
Bed x: 144.00 y: 50.00 z: 0.12
Bed x: 145.00 y: 100.00 z: 0.28
Bed x: 88.00 y: 100.00 z: 0.08
Bed x: 31.00 y: 100.00 z: -0.24
Bed x: 30.00 y: 150.00 z: -0.27
Bed x: 87.00 y: 150.00 z: -0.01
Bed x: 144.00 y: 150.00 z: 0.37
Eqn coefficients: a: 0.00 b: 0.00 d: -0.44
planeNormal x: -0.00 y: -0.00 z: 1.00
echo:endstops hit:  Z:0.37
echo:ok 999.99,0.00,4.31,-0.00,1000.00,0.71,-4.31,-0.71,999.99.

Bed x: 30.00 y: 50.00 z: -0.20
Bed x: 87.00 y: 50.00 z: 0.01
Bed x: 144.00 y: 50.00 z: 0.11
Bed x: 145.00 y: 100.00 z: 0.26
Bed x: 88.00 y: 100.00 z: 0.05
Bed x: 31.00 y: 100.00 z: -0.22
Bed x: 30.00 y: 150.00 z: -0.29
Bed x: 87.00 y: 150.00 z: 0.11
Bed x: 144.00 y: 150.00 z: 0.37
Eqn coefficients: a: 0.00 b: 0.00 d: -0.44
planeNormal x: -0.00 y: -0.00 z: 1.00
echo:endstops hit:  Z:0.37
echo:ok 999.99,0.00,4.27,-0.00,1000.00,0.89,-4.27,-0.89,999.99.

ABLG Glass plate 2, 2 clamps

Bed x: 30.00 y: 50.00 z: -0.29
Bed x: 87.00 y: 50.00 z: -0.01
Bed x: 144.00 y: 50.00 z: 0.13
Bed x: 145.00 y: 100.00 z: 0.31
Bed x: 88.00 y: 100.00 z: 0.10
Bed x: 31.00 y: 100.00 z: -0.28
Bed x: 30.00 y: 150.00 z: -0.38
Bed x: 87.00 y: 150.00 z: 0.07
Bed x: 144.00 y: 150.00 z: 0.38
Eqn coefficients: a: 0.01 b: 0.00 d: -0.52
planeNormal x: -0.01 y: -0.00 z: 1.00
echo:endstops hit:  Z:0.38
echo:ok 999.99,0.00,5.15,-0.00,1000.00,0.78,-5.15,-0.78,999.99.

Bed x: 30.00 y: 50.00 z: -0.30
Bed x: 87.00 y: 50.00 z: -0.02
Bed x: 144.00 y: 50.00 z: 0.11
Bed x: 145.00 y: 100.00 z: 0.29
Bed x: 88.00 y: 100.00 z: 0.04
Bed x: 31.00 y: 100.00 z: -0.32
Bed x: 30.00 y: 150.00 z: -0.44
Bed x: 87.00 y: 150.00 z: 0.04
Bed x: 144.00 y: 150.00 z: 0.34
Eqn coefficients: a: 0.01 b: 0.00 d: -0.54
planeNormal x: -0.01 y: -0.00 z: 1.00
echo:endstops hit:  Z:0.35
echo:ok 999.99,0.00,5.29,-0.00,1000.00,0.55,-5.29,-0.55,999.99.

ABLG Glass plate 1, 4 clamps

Bed x: 30.00 y: 50.00 z: -0.58
Bed x: 87.00 y: 50.00 z: -0.15
Bed x: 144.00 y: 50.00 z: 0.22
Bed x: 145.00 y: 100.00 z: 0.45
Bed x: 88.00 y: 100.00 z: -0.05
Bed x: 31.00 y: 100.00 z: -0.41
Bed x: 30.00 y: 150.00 z: -0.42
Bed x: 87.00 y: 150.00 z: 0.12
Bed x: 144.00 y: 150.00 z: 0.60
Eqn coefficients: a: 0.01 b: 0.00 d: -0.98
planeNormal x: -0.01 y: -0.00 z: 1.00
echo:endstops hit:  Z:0.60
echo:ok 999.97,0.00,7.89,-0.02,1000.00,2.68,-7.89,-2.68,999.97.

Bed x: 30.00 y: 50.00 z: -0.60
Bed x: 87.00 y: 50.00 z: -0.14
Bed x: 144.00 y: 50.00 z: 0.21
Bed x: 145.00 y: 100.00 z: 0.47
Bed x: 88.00 y: 100.00 z: 0.02
Bed x: 31.00 y: 100.00 z: -0.39
Bed x: 30.00 y: 150.00 z: -0.39
Bed x: 87.00 y: 150.00 z: 0.17
Bed x: 144.00 y: 150.00 z: 0.62
Eqn coefficients: a: 0.01 b: 0.00 d: -1.00
planeNormal x: -0.01 y: -0.00 z: 1.00
echo:endstops hit:  Z:0.62
echo:ok 999.97,0.00,7.86,-0.02,1000.00,3.09,-7.86,-3.09,999.96.

ABLG Glass plate 1, 2 clamps

Bed x: 30.00 y: 50.00 z: -0.25
Bed x: 87.00 y: 50.00 z: 0.02
Bed x: 144.00 y: 50.00 z: 0.13
Bed x: 145.00 y: 100.00 z: 0.34
Bed x: 88.00 y: 100.00 z: 0.12
Bed x: 31.00 y: 100.00 z: -0.21
Bed x: 30.00 y: 150.00 z: -0.30
Bed x: 87.00 y: 150.00 z: 0.15
Bed x: 144.00 y: 150.00 z: 0.48
Eqn coefficients: a: 0.01 b: 0.00 d: -0.53
planeNormal x: -0.01 y: -0.00 z: 1.00
echo:endstops hit:  Z:0.48
echo:ok 999.99,0.00,5.01,-0.01,1000.00,1.42,-5.01,-1.42,999.99.

Bed x: 30.00 y: 50.00 z: -0.25
Bed x: 87.00 y: 50.00 z: -0.01
Bed x: 144.00 y: 50.00 z: 0.13
Bed x: 145.00 y: 100.00 z: 0.31
Bed x: 88.00 y: 100.00 z: 0.07
Bed x: 31.00 y: 100.00 z: -0.25
Bed x: 30.00 y: 150.00 z: -0.35
Bed x: 87.00 y: 150.00 z: 0.10
Bed x: 144.00 y: 150.00 z: 0.42
Eqn coefficients: a: 0.00 b: 0.00 d: -0.52
planeNormal x: -0.00 y: -0.00 z: 1.00
echo:endstops hit:  Z:0.42
echo:ok 999.99,0.00,4.98,-0.01,1000.00,1.02,-4.98,-1.02,999.99.

**Comment by imqqmi** [comment]
I've found another weak point in my hardware: the tube connecting the Z motor to the rod. I've tried tightening it with zip ties but there was still too much play to my taste and replaced it with an aluminium coupler. These changes have the negative effect that the wobble of the rod translates to the head but I'll fix that another day. The grid probing is now even more accurate with a deviation of 0.01mm down from 0.03 to 0.05. The values I now get are much different, much farther apart. 

I found out the tube was a weak link in the chain because I wanted to try full step instead of 1/16th step. The motor turned really slowly (seemed like it ignored the jumper change on the RAMPS board). But I did notice the tube slipping at that slow turning rate. 

So another day of testing how this will affect my prints. I'll keep you posted.

**Comment by imqqmi** [comment]
Printing with 3P ABL or ABLG height of a 25mm object printed 24.5. And that there no difference between the two, which is good! I'm have suspicion that the z rods may not be exactly 0.8mm. I've found a source for trapezoidal rods, probably going to be 8mm. I'll have to do some mods to the prusa i3 but I think/hope this should fix the errors once and for all.

**Comment by boelle** [comment]
Is this one still a bug or has the issue been solved?

**Comment by imqqmi** [comment]
If no one else can reproduce it then it must be my printer hardware that's to blame. Still, there might be a bug that lowers the height if the print bed isn't very flat. That's something that should be fixed first in any case of course.

Anyway if anyone is willing to look into it, they can use the bed level probe values I've posted earlier and plug them into ABLG and see what the output is when printing a single walled test cube. 

The steps I've taken to resolve my hardware was:
1. Fix the play/backlash in the nut in the x motor part that's on the Z axis (may introduce wobble if the rod isn't straight, stainless steel/ACME/trapezoidal rods recommended)
2. Get rid of the PTFE tube that connects the Z motor to the Z rod. Mine slipped, and when fixed with zip ties torsion on the tube gave backlas/play. Replace with aluminium coupler. These #1 and #2 fixes increased consistency and accuracy of the probing ten fold.
3. Use the ABLG probe readings in pronterface to get an idea how unlevel the bed is and try to level it. use only two clamps to clamp on the glass or metal plate as four will bend it if the backing isn't flat (by backing I mean heatbed or any other plate behind the printbed).

My setup doesn't experience the problem anymore in any case.

**Comment by wgm4321** [comment]
I don't believe the behavior I have seen is a bug or even a vaild use case (multiple G29s)

**Comment by nicolas-rambaud** [comment]
I’m glad you could find  a proper solution to this ;)



*De :* imqqmi [mailto:notifications@github.com]
*Envoyé :* mercredi 31 décembre 2014 19:13
*À :* ErikZalm/Marlin
*Cc :* Nicolas Rambaud
*Objet :* Re: [Marlin] Auto bed leveling grid on/off Z height difference
(#1244)



Yes! I've done another dry run and the results are consistent! Thanks a lot
guys for putting me on the right track! I've been at this for days!
@brainscan <https://github.com/brainscan>: Agreed! A great start of the new
year! Oh man, seems like everything that could go wrong with this printer
went wrong. Every step in the right direction had to be won. But I'm loving
the experience on the whole, learned a lot!

—
Reply to this email directly or view it on GitHub
<https://github.com/ErikZalm/Marlin/issues/1244#issuecomment-68459099>.[image:
Image supprimée par l'expéditeur.]

**Comment by imqqmi** [comment]
Thanks @nicolas-rambaud! I've bought stainless steel M5 rods which look much better than the zinc alloys. Also a lot straighter so I have good hope it will eliminate the wobble as well. I couldn't get hold of trapezoidal rods of 5 or 8mm, only 10mm, which may be a bit too heavy for the stepper motors, and more difficult to find couplers and adjusting the parts. The M5 stainless steel rods are a drop in replacement and should be easy to do. 

Cheers!

**Comment by wgm4321** [comment]
Can this be closed?   I believe that both of us that reported behaviors as a bug now agree that we aren't seeing a bug.

**Comment by imqqmi** [comment]
@wgm4321 agreed, I will close it. 

As promised I'd report back when I've replaced the Z rods. I've compared the rods and can see a slight offset in the thread. The cheap ones threads travel a shorter distance per turn compared to the stainless steel one. 

The height is better. With Z_STEPS at 4000 the height is 24.65. It's pretty close to the 24.85mm it should've been. I can probably compensate it with Z offset. 

I've scanned the rods. The cheap rod is a factor 0.9985375 too small per turn while the stainless steel is 1.001975 off of 1. So in theory I should put in a smaller number in the Z_STEPS. 1/1.001975 * 4000 = 3995.2. 

I'm now at 4012 Z_STEPS printing on target at 24.85mm (asked for, and printed). 
Here's the scan if you're interested, the dots represent 10 turns in the thread for easy counting and measuring, scanned at 1200 dpi. To convert from pixels to mm use pixels/dpi*2.54:
![Zinc VS stainless steel rods]](http://www.jmbfoto.nl/WireEndsZincVSStainlessSteel.jpg)
